services/venice-controller/src/main/java/com/linkedin/venice/controller/UnusedValueSchemaCleanupService.java

```diff
@@ -0,0 +1,151 @@
+      VeniceParentHelixAdmin parentHelixAdmin) {
+    this.multiClusterConfig = multiClusterConfig;
+    this.admin = admin;
+    this.scheduleIntervalMinutes = multiClusterConfig.getUnusedSchemaCleanupIntervalMinutes();
+    this.minSchemaCountToKeep = multiClusterConfig.getMinSchemaCountToKeep();
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+  }
+
+  private Runnable getRunnableForSchemaCleanup() {
+    return () -> {
+      if (stop) {
+        return;
+      }
+      for (String clusterName: multiClusterConfig.getClusters()) {
+        boolean cleanupEnabled =
+            multiClusterConfig.getControllerConfig(clusterName).isUnusedValueSchemaCleanupServiceEnabled();
+        if (!cleanupEnabled) {
+          continue;
+        }
+        // Get all stores for current cluster
+        List<Store> stores = admin.getAllStores(clusterName);
+        for (Store store: stores) {
+          String storeName = store.getName();
+          // Remove schema only for batch stores
+          if (isHybridStore(store) || VeniceSystemStoreUtils.isSystemStore(storeName)) {
+            continue;
+          }
+          Collection<SchemaEntry> allSchemas = veniceParentHelixAdmin.getValueSchemas(clusterName, storeName);
+
+          if (allSchemas.size() < minSchemaCountToKeep) {
+            continue;
+          }
+
+          Set<Integer> inUseValueSchemaIds = veniceParentHelixAdmin.getInUseValueSchemaIds(clusterName, storeName);
+
+          // if any of the child colo is unreachable, skip deletion.
+          if (inUseValueSchemaIds.isEmpty()) {
+            LOGGER.warn("Could not find in-use value schemas for store {}", storeName);
+            continue;
+          }
+
+          Set<Integer> schemasToDelete = findSchemaIdsToDelete(
+              allSchemas,
+              store,
+              admin.getHelixVeniceClusterResources(clusterName).getSchemaRepository(),
+              inUseValueSchemaIds);
+
+          if (!schemasToDelete.isEmpty()) {
+            // delete from child colos
+            veniceParentHelixAdmin.deleteValueSchemas(clusterName, store.getName(), schemasToDelete);
+            // delete from the parent colo
+            admin.deleteValueSchemas(clusterName, store.getName(), schemasToDelete);
+          }
+        }
+      }
+    };
+  }
+
+  private boolean isHybridStore(Store store) {
+    if (store.getHybridStoreConfig() != null) {
+      return true;
+    }
+    // check future/backup versions
+    for (Version version: store.getVersions()) {
+      if (version.getHybridStoreConfig() != null) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  Set<Integer> findSchemaIdsToDelete(
+      Collection<SchemaEntry> allSchemas,
+      Store store,
+      ReadWriteSchemaRepository schemaRepository,
+      Set<Integer> inUseValueSchemaIds) {
+    Set<Integer> schemasToDelete = new HashSet<>();
+
+    // assumes `getValueSchemas` returns ascending schema ids so that the older schemas are deleted first
```

Comment:
Let us not assume this order, and can we sort it before the cleanup?

Reply:
sure